flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java

```diff
@@ -127,7 +155,28 @@ public void increaseResourceRequirementsBy(ResourceCounter increment) {
         }
         totalResourceRequirements = totalResourceRequirements.add(increment);
 
+        if (slotRequestMaxInterval == null) {
+            declareResourceRequirements();
+            return;
+        }
+
+        Preconditions.checkNotNull(componentMainThreadExecutor);
+        if (slotRequestMaxIntervalTimeoutCheckFuture != null) {
+            slotRequestMaxIntervalTimeoutCheckFuture.cancel(true);
+        }
+        slotRequestMaxIntervalTimeoutCheckFuture =
+                componentMainThreadExecutor.schedule(
+                        this::checkSlotRequestMaxIntervalTimeout,
+                        slotRequestMaxInterval.toMilliseconds(),
+                        TimeUnit.MILLISECONDS);
+    }
+
+    private void checkSlotRequestMaxIntervalTimeout() {
+        if (componentMainThreadExecutor == null || slotRequestMaxInterval == null) {
+            return;
+        }
         declareResourceRequirements();
+        slotRequestMaxIntervalTimeoutCheckFuture = null;
```

Comment:
Do we ensure that all the modification of this variable are happened in the same thread?

Reply:
Yes, here is an illegal assignment with risk.
IIUC, we only need to close the current `future` that's cancelable when increasing requirements. Otherwise, we can use the result of scheduled task to make assignment, please correct me if i'm wrong.